src/main/java/com/raoulvdberge/refinedstorage/apiimpl/autocrafting/CraftingPattern.java

```diff
@@ -273,7 +273,28 @@ public String toString() {
             ", byproducts=" + byproducts +
             '}';
     }
-
+    
+    @Override
+    public boolean equals (Object obj) {
+    	return this==obj || (obj instanceof ICraftingPattern && this.alike((ICraftingPattern)obj));
+    }
+    
+    Integer hashCode;
```

Comment:
Would mark this private and any reason for it being the boxed `Integer` instead of the primitive `int`?

Reply:
I think he used boxing for caching the hash (if null compute hash).

Either way, this field has to be moved to the top like other fields and indeed needs a visibility modifier (private).

Reply:
Might want to make it `Integer hashCode = null;` and maybe add a little like saying this is for caching reasons. Or just call it `hashCodeCached`